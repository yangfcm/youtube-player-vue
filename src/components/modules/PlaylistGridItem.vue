<template>
  <div class="ui card centered" style="width: 100%; height: 100%; margin: 0;">
    <router-link :to="`/playlist/${playlist.id}`" class="image">
      <img :src="playlist.snippet.thumbnails.medium.url" :alt="playlist.snippet.title" />
    </router-link>
    <div class="content">
      <router-link :to="`/playlist/${playlist.id}`" class="header">{{ playlist.snippet.title}}</router-link>
      <div class="ui blue label tiny" style="margin-left: 0; margin-top: 3px;">
        {{ playlist.contentDetails.itemCount }}
        <span v-if="playlist.contentDetails.itemCount">videos</span>
        <span v-else>video</span>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: ["playlist"],
};
</script>

<style scoped>
</style>