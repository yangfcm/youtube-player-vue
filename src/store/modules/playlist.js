import axios from "../../settings";
import {
  FETCH_PLAY_LIST,
  FETCH_PLAY_LIST_DETAIL,
  CATCH_ERROR,
  FETCH_MY_PLAY_LIST,
} from "../types";
import parseError from "../helpers/parseError";

const state = {
  playlist: null,
  myPlaylist: null,
  playlistDetail: null,
  playlistError: null,
};

const getters = {
  playlist: (state) => state.playlist,
  myPlaylist: (state) => state.myPlaylist,
  playlistDetail: (state) => state.playlistDetail,
  playlistError: (state) =>
    parseError(state.playlistError, "Failed to fetch playlist videos"),
};

const mutations = {
  FETCH_PLAY_LIST: (state, payload) => {
    state.playlist = payload;
    state.playlistError = null;
  },
  FETCH_MY_PLAY_LIST: (state, payload) => {
    state.myPlaylist = payload;
    state.playlistError = null;
  },
  FETCH_PLAY_LIST_DETAIL: (state, payload) => {
    state.playlistDetail = payload;
    state.playlistError = null;
  },
  CATCH_ERROR: (state, payload) => {
    state.playlistError = payload;
  },
};

const actions = {
  /** Fetch the playlist of authed user */
  fetchMyPlaylist: async (context, pageToken = null) => {
    const accessToken = localStorage.getItem("access_token");
    try {
      const response = await axios.get("/playlists", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          part: "snippet,contentDetails",
          maxResults: 12,
          pageToken,
          mine: true,
        },
      });
      context.commit(FETCH_MY_PLAY_LIST, response.data);
    } catch (err) {
      context.commit(CATCH_ERROR, err);
    }
  },

  /** Fetch the playlist that belongs to a channel */
  fetchPlaylist: async (context, channelId, pageToken = null) => {
    try {
      const response = await axios.get("/playlists", {
        params: {
          ...axios.defaults.params,
          part: "snippet,contentDetails",
          maxResults: 12,
          channelId,
          pageToken,
        },
      });
      context.commit(FETCH_PLAY_LIST, response.data);
    } catch (err) {
      context.commit(CATCH_ERROR, err);
    }
  },

  /** Fetch the videos in a playlist */
  fetchPlaylistDetail: async (context, playlistId, pageToken = null) => {
    try {
      const response = await axios.get("/playlistItems", {
        params: {
          ...axios.defaults.params,
          part: "snippet,contentDetails",
          maxResults: 10,
          playlistId,
          pageToken,
        },
      });
      context.commit(FETCH_PLAY_LIST_DETAIL, response.data);
    } catch (err) {
      context.commit(CATCH_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
