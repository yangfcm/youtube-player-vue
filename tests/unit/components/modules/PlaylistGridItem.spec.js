import { shallowMount, RouterLinkStub } from "@vue/test-utils";
import PlaylistGridItem from "@/components/modules/PlaylistGridItem.vue";
import { playlistResponse } from "../../store/fixtures/playlist";

describe("Test PlaylistGridItem component", () => {
  let wrapper;
  const playlist = playlistResponse.items[0];
  beforeEach(() => {
    wrapper = shallowMount(PlaylistGridItem, {
      propsData: {
        playlist,
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct content", () => {
    expect(wrapper.find("img").attributes("src")).toBe(
      playlist.snippet.thumbnails.medium.url
    );
    expect(wrapper.find(".header").text()).toBe(playlist.snippet.title);
    expect(wrapper.find("a").props().to).toBe(`/playlist/${playlist.id}`);
    expect(wrapper.find("div.tiny.label").text()).toContain(
      playlist.contentDetails.itemCount
    );
  });
});
