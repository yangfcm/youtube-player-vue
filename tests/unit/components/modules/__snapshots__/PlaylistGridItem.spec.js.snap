// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PlaylistGridItem component should match snapshot 1`] = `
<div class="ui card centered" style="width: 100%; height: 100%; margin: 0px;"><a class="image"><img src="https://i.ytimg.com/vi/f0jETLKTW7w/mqdefault.jpg" alt="TED Connects: Community and Hope"></a>
  <div class="content"><a class="header">TED Connects: Community and Hope</a>
    <div class="ui blue label tiny" style="margin-left: 0px; margin-top: 3px;">
      15
      <span>videos</span></div>
  </div>
</div>
`;
